File: api/src/adapters/keyvault.ts
import { SecretClient } from '@azure/keyvault-secrets';
import { ClientSecretCredential } from '@azure/identity';
import config from '../lib/config';

type SecretsKeys =
  | 'dummySecret'
  | 'AIConnectionString'
  | 'blobStorageConnectionString'
  | 'imageContainerName'
  | 'dbHost'
  | 'dbPassword'
  | 'dbUsername'
  | 'dbName'
  | 'dbConnectionString';

const secrets = new Map<SecretsKeys, string | undefined>();

const getSecrets = async () => {
  if (secrets.size !== 0) return secrets;

  const keyVaultName = config.azure.keyvault.name;
  const clientSecret = config.azure.keyvault.clientSecret;
  const clientId = config.azure.keyvault.clientId;
  const tenantId = config.azure.keyvault.tenantId;

  if (keyVaultName == null) {
    throw new Error('config.azure.keyvault.name is empty');
  }
  if (clientId == null) {
    throw new Error('config.azure.keyvault.clientId is empty');
  }
  if (clientSecret == null) {
    throw new Error('config.azure.keyvault.clientSecret is empty');
  }
  if (tenantId == null) {
    throw new Error('config.azure.keyvault.tenantId is empty');
  }

  const credential = new ClientSecretCredential(
    tenantId,
    clientId,
    clientSecret,
  );
  const url = 'https://' + keyVaultName + '.vault.azure.net';

  const client = new SecretClient(url, credential);

  for await (const secretProperties of client.listPropertiesOfSecrets()) {
    const secret = await client.getSecret(secretProperties.name);
    secrets.set(secretProperties.name as SecretsKeys, secret.value);
  }

  return secrets;
};

export default getSecrets;

File: backend/adapters/keyvault.py
from . import config
from azure.identity import ClientSecretCredential
from azure.keyvault.secrets import SecretClient

secrets = {}

def get_secrets():
    if len(secrets) != 0:
        return secrets

    conf = config.load()
    azure_key_vault = conf['defaults']['azure']['keyvault']

    keyvault_name = azure_key_vault['name']
    client_secret = azure_key_vault['clientSecret']
    client_id = azure_key_vault['clientId']
    tenant_id = azure_key_vault['tenantId']

    credential = ClientSecretCredential(
            tenant_id=tenant_id,
            client_secret=client_secret,
            client_id=client_id,
    )

    url = 'https://' + keyvault_name + '.vault.azure.net'
    client = SecretClient(vault_url=url, credential=credential)

    for secret_properties in client.list_properties_of_secrets():
        name = secret_properties.name if secret_properties.name is not None else ''
        if len(name) == 0:
            continue
        secret = client.get_secret(name=name)
        secrets[secret_properties.name] = secret.value
    return secrets

